### PR TITLE
Add IMDWorkspace isHistogramData function

### LINKED ISSUE #16977: Add a function for check whether a IMDWorkspace is actually a IMDHistoWorkspace
There is currently (as far as I can see) no easy way of checking if a IMDWorkspace is a IMDHistoWorkspace without using a dynamic cast. This leads to selection statements like the following in SliceViewer.cpp:

```cpp
if(boost::dynamic_pointer_cast<IMDHistoWorkspace>(m_ws) {
  ...
}
``` 
The goal of this issue is to:
 - Add a function to provide this check
 - Replace code like the line above with the new function.

### PR DESCRIPTION
**Pull request by samueljackson92**
This PR is a spin off from PR #16922. Currently the only way to tell if a workspace was a histogram workspace was to dynamically cast it. 

The aim of this PR was to create a function that would allow us to avoid doing that by modifying IMDWorkspace. As we already have a function called isHistogramData() one way to do this is to move this one level up and implement it for MDEventWorkspace and MDHistoWorkspace. 

**To test:**
There's not much to test other than that the unit tests & system tests pass and that the code is good.

Fixes #16977.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc [hunk 1]
```diff
@@ -1011,6 +1011,13 @@ TMDE(void MDEventWorkspace)::setDisplayNormalization(
 }
 
 /**
+  Return if the workspace is a histogram.
+*/
+TMDE(bool MDEventWorkspace)::isHistogramData() const {
+  return false;
```
Comment (SimonHeybrock) [inline Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc]: Why? Isn't an MD workspace always histogram data, in the sense of the `MatrixWorkspace::isHistogramData()` method?

**Comment by SimonHeybrock** [comment]
@samueljackson92 If I understand the issue description correctly the method you need is *NOT* equivalent to `MatrixWorkspace::isHistogramData()`. The latter indicates whether the X data corresponds to bin edges or the data points (same length of X and Y vectors). In that sense `EventWorkspace` is also histogram data and so is, I think, every MD workspace including `MDEventWorkspace`. The grid boxes correspond to bin edges.

**Comment by samueljackson92** [comment]
@SimonHeybrock My mistake. I've misunderstood the meaning of this method. You are correct, this should be separate method for differentiating MDEvent from MDHisto and that is not equivalent.

File: Framework/PythonInterface/mantid/api/src/Exports/MatrixWorkspace.cpp [hunk 2]
```diff
@@ -230,8 +230,6 @@ void export_MatrixWorkspace() {
       .def("getAxis", &MatrixWorkspace::getAxis,
            (arg("self"), arg("axis_index")), return_internal_reference<>(),
            "Get a pointer to a workspace axis")
-      .def("isHistogramData", &MatrixWorkspace::isHistogramData, arg("self"),
-           "Returns True if this is considered to be binned data.")
```
Comment (SimonHeybrock) [inline Framework/PythonInterface/mantid/api/src/Exports/MatrixWorkspace.cpp]: I think this change should also be reverted?

File: MantidQt/API/test/SignalRangeTest.h [hunk 3]
```diff
@@ -44,6 +44,7 @@ class SignalRangeTest : public CxxTest::TestSuite {
     MOCK_METHOD0(clearMDMasking, void());
     MOCK_CONST_METHOD0(getSpecialCoordinateSystem,
                        Mantid::Kernel::SpecialCoordinateSystem());
+    MOCK_CONST_METHOD0(isHistogramData, bool());
```
Comment (SimonHeybrock) [inline MantidQt/API/test/SignalRangeTest.h]: Is this correct?

**Comment by SimonHeybrock** [comment]
Changes look ok now. This helps to avoid some `dynamic_casts`. If Jenkins gives his ok I would say this is ready for :shipit:.

**Pull request merged**